pkg/transformers/namereference.go

```diff
@@ -106,6 +106,97 @@ func (o *nameReferenceTransformer) Transform(m resmap.ResMap) error {
 	return nil
 }
 
+// utility function to replace a simple string by the new name
+func (o *nameReferenceTransformer) selectReferral(
+	oldName string,
+	referrer *resource.Resource,
+	target gvk.Gvk,
+	referralCandidates resmap.ResMap,
+	referralCandidateSubset []*resource.Resource) (interface{}, interface{}, error) {
+
+	for _, res := range referralCandidateSubset {
+		id := res.OrgId()
+		if id.IsSelected(&target) && res.GetOriginalName() == oldName {
+			matches := referralCandidates.GetMatchingResourcesByOriginalId(id.Equals)
+			// If there's more than one match, there's no way
+			// to know which one to pick, so emit error.
+			if len(matches) > 1 {
+				return nil, nil, fmt.Errorf(
+					"multiple matches for %s:\n  %v",
+					id, getIds(matches))
+			}
+			// In the resource, note that it is referenced
+			// by the referrer.
+			res.AppendRefBy(referrer.CurId())
+			// Return transformed name of the object,
+			// complete with prefixes, hashes, etc.
+			return res.GetName(), res.GetNamespace(), nil
+		}
+	}
+
+	return oldName, nil, nil
+}
+
+// utility function to replace a simple string by the new name
+func (o *nameReferenceTransformer) mapSimpleNameField(
+	oldName string,
+	referrer *resource.Resource,
+	target gvk.Gvk,
+	referralCandidates resmap.ResMap,
+	referralCandidateSubset []*resource.Resource) (interface{}, error) {
```

Comment:
You can remove `referralCandidateSubset` since it can be get from `referralCandiates.Resources()`

Reply:
We can not remove it. The whole handling of namespaces relies on it:. When we have to deal with the same name coming from different namespace with use the original namespace to filter the candidate: https://github.com/kubernetes-sigs/kustomize/blob/579995dc8a1e3e1df267f5c57f9d54377200f6c2/pkg/transformers/namereference.go#L180

Reply:
I understand that it can't be removed from `selectReferral` function based on current logic. However, when `mapSimpleNameField` is called, this resource list is always the resources from the passed resmap.

One more thing, in `mapNameandNSstruct`, is that possible to filter the resmap with the namespace? So that the resmap passed to `selectReferral` all match with that namespace.

Reply:
Added comment to explain the reasoning: https://github.com/kubernetes-sigs/kustomize/pull/1325/commits/9b40f8ab474007a4f2358586bc82460b4bcac577

Reply:
@Liujingfang1 I don't think we should filter in the getNameAndNSStruct because would would most likely hide conflicts instead of the detecting them in the getSimpleNameField. This is the only reason why the resMap is passed as parameter.